src/NUnitFramework/framework/Api/NUnitTestAssemblyRunner.cs

```diff
@@ -373,12 +373,13 @@ private void OnRunCompleted(object sender, EventArgs e)
         private int CountTestCases(ITest test, ITestFilter filter)
         {
             if (!test.IsSuite)
-                return 1;
+                return filter.Pass(test) ? 1: 0;
```

Comment:
Pass would walk both parents and children of the node. This optimization makes it so we only call Pass on the leaf nodes avoiding repeated enumeration and evaluation of descendants.

Reply:
@jnm2 I haven't reviewed the PR but caught your mention of me and wanted to reply.

Yes, it does apply to all nodes. It probably not clear these days because the docs say that the filter applies to "tests". People accustomed to the Microsoft runners tend to think that a "test" means what we call a "test case".

The existing logic relies on the fact that the check is made at the top level only. Since we never call CountTestCases on a node we don't want to have included in the count, the filter check seems redundant... that is, __unless__ the other changes in this PR cause this private member to be called in some recursive way.

Although I haven't reviewed this, note that the purpose of the change is to make a particular type of filter -- a large number of test name filters joined by or -- more efficient. Of course, that's a filter, which no one familiar with how NUnit works should ever construct, but it's also how Microsoft constructs filters. They do that precisely because they want it to work with every runner possible. I've suggested before that the vs adapter is the place to make such a change because the adapter is the sole piece of code, which combines knowledge of how NUnit works with knowledge of how Microsoft works.

That said, it's true that the filtering logic could be more efficient. :smile:

Reply:
Wondering how you can filter looking only at leaf nodes and still handle Categories correctly.  That is, you can deduce the name of all the suites that contain a test case from the test case itself, but you can't do that with categories.

Reply:
>Wondering how you can filter looking only at leaf nodes and still handle Categories correctly.

`.Pass()` walks parents of the test as well as the node itself and the descendants. So calling it on the leaf node would check it's parents as well.

Reply:
>If we pass in a negative filter, then we will only match the current ITest and the parents, so I that setting this change will make the performance worse as far as I can tell (as we could short-circuit the checks if some of the parents fail the check).

Correct, but re-walking parents is comparatively cheap (there are usually few levels) to walking the descendants. This seems like a tradeoff worth making.

Reply:
That's OK then. I was going by your description of the change... i.e. that you would only look at leaf nodes. You're still looking at them, but indirectly through `Pass`.

I remember we had a reason to look at both ancestors and descendants of each node in the original V2 code. We wanted to avoid doing both but for some reason couldn't. Wish I could remember the reason!!! In any case, if you can make all tests pass, that reason might be obsolete. Still, if I can figure out what it was, I'll post here again.

Reply:
> Although I haven't reviewed this, note that the purpose of the change is to make a particular type of filter -- a large number of test name filters joined by or -- more efficient.

As far as I can tell, the changes made in this file and in all the other files in this PR, except for OrFilter.cs, improves the general performance (of filters) and are not related to the particular filter created by vsTest.

> In any case, if you can make all tests pass, that reason might be obsolete

I think this change is sound, but making all test pass is a very low bar at the moment, as we only have one test that tests `CountTestCases` with a non empty filter. So I would be much more confident if we had some more tests of this (I would also like to test this some more manually). Unfortunately, my entire weekend is packed, so I don't think I'll have much time for this before next week.

Reply:
>  So I would be much more confident if we had some more tests of this

I'll add some.

Reply:
>So I would be much more confident if we had some more tests of this

Added CountTestCases tests that match the assembly, category, fixture, and individual tests.